armi/reactor/compositeList.py

```diff
@@ -0,0 +1,246 @@
+"""Collection of composites with a 1-D grid"""
+
+from abc import ABC, ABCMeta, abstractmethod
+from typing import Optional, Iterator, Tuple
+
+from numpy import empty, ndarray
+
+from armi import runLog
+from armi.reactor import flags
+from armi.reactor.grids import Grid, axialUnitGrid
+from armi.reactor.composites import Composite, CompositeModelType
+
+
+class AbstractCompositeMetaclass(ABCMeta, CompositeModelType):
+    """Metaclass that allows for abstract composites"""
+
+
+class AbstractComposite(ABC, Composite, metaclass=AbstractCompositeMetaclass):
+    """Composite that can have abstract methods
+
+    Useful for enforcing interfaces across generic subclasses
+    """
+
+
+class CompositeWithHeight(AbstractComposite):
+    """Interface for a composite with a height.
+
+    Like a block but more general
+
+    """
+
+    @abstractmethod
+    def getHeight(self) -> float:
+        """Return the height of this object"""
+
+
+class CompositeList(CompositeWithHeight):
+    """Axial collection of composites
+
+    Many things in an ARMI :class:`~armi.reactor.reactors.Reactor`
+    can be seen as things stacked on things. An
+    :class:`~armi.reactor.assemblies.Assemblies` is a 1-D stack of
+    :class:`~armi.reactor.blocks.Block` instances, each with an axial
+    location and height. Fuel pins are another concept that would benefit
+    from a 1-D stack of :class:`~armi.reactor.composites.Composite` like things.
+
+    This class seeks to separate out storing and maintaing the axial mesh
+    of things (iteration, item access, mesh updating, etc.) from
+    assembly-specific tasks (moving assemblies within the reactor,
+    looking for coolant channels, etc.)
+
+    """
+
+    spatialGrid: Optional[Grid]
+
+    def __init__(self, name):
+        super().__init__(name)
```

Comment:
Style points thing.

For reasons that pre-date my involvement in this project, we have [a style guide in the docs](https://terrapower.github.io/armi/developer/standards_and_practices.html?highlight=super#general-do-s-and-don-ts) that says to not use `super()`, in favor of more explicit calls to the super class `__init__()` function.

Just to be entirely clear, we use `super()` all over ARMI, and no one has bothered to fix it yet.  So, it's not mission-critical. But probably we should at least leave the `Assembly` one as-is.

Reply:
Roger. We will need to change the assembly `__init__` to not point to `Composite.__init__` instead the parent class, but I think the spirit of the request would still be preserved

Reply:
How do we feel about this in light of #1155?